### ISSUE TITLE
`req` app: `-addext` option does not combine with other options adding X.509v3 extensions

### ISSUE THREAD

**Issue by stefanholek** [openssl/openssl#11169]
I was hoping to use openssl req -addext to add subjectAltNames to my CSRs but no dice.

```
openssl version
OpenSSL 1.1.1d  10 Sep 2019
```

Given a config file for TLS server requests (server.conf):
```
[ req ]
default_bits            = 2048                  # RSA key size
encrypt_key             = no                    # Protect private key
default_md              = sha256                # MD to use
prompt                  = yes                   # Prompt for DN
distinguished_name      = server_dn             # DN template
req_extensions          = server_reqext         # Desired extensions

[ server_dn ]
countryName             = "1. Country Name (2 letters) (eg, US)       "
countryName_max         = 2
stateOrProvinceName     = "2. State or Province Name   (eg, region)   "
localityName            = "3. Locality Name            (eg, city)     "
organizationName        = "4. Organization Name        (eg, company)  "
organizationalUnitName  = "5. Organizational Unit Name (eg, section)  "
commonName              = "6. Common Name              (eg, FQDN)     "
commonName_max          = 64

[ server_reqext ]
keyUsage                = critical,digitalSignature,keyEncipherment
extendedKeyUsage        = serverAuth,clientAuth
subjectKeyIdentifier    = hash
```

I create a CSR with:
```
openssl req —new -config server.conf -out foo.csr -keyout foo.key

openssl req -in foo.csr -noout -text -reqopt no_pubkey,no_sigdump
Certificate Request:
    Data:
        Version: 1 (0x0)
        Subject: C = US, O = Foo Inc, CN = www.foo.com
        Attributes:
        Requested Extensions:
            X509v3 Key Usage: critical
                Digital Signature, Key Encipherment
            X509v3 Extended Key Usage: 
                TLS Web Server Authentication, TLS Web Client Authentication
            X509v3 Subject Key Identifier: 
                3A:5F:6C:71:A8:F3:4C:48:F5:8B:41:25:22:7A:16:8C:05:2F:9D:DB
```

Fine so far. Now to add a subjectAltName extension:
```
openssl req —new -config server.conf -out foo.csr -keyout foo.key -addext subjectAltName=DNS:www.foo.com

openssl req -in foo.csr -noout -text -reqopt no_pubkey,no_sigdump
Certificate Request:
    Data:
        Version: 1 (0x0)
        Subject: C = US, O = Foo Inc, CN = www.foo.com
        Attributes:
        Requested Extensions:
            X509v3 Subject Alternative Name: 
                DNS:www.foo.com
```

As you can see all extensions from the config file have disappeared, which is highly unexpected given the option is called -addext. :-)

**Comment by bmurray** [comment]
I believe this is related to #10728

After some investigating in the head, I don't believe it is adding extensions correctly, but instead replacing them and is somehow bound by the size of the previous extensions added. I'm going to keep digging. addext is a pretty new, but also extremely useful feature.

**Comment by kwaclaw** [comment]
Bug still exists in OpenSSL 3.0. -addext would have been useful.

**Comment by nhorman** [comment]
@kwaclaw would you mind testing with 3.3 to confirm the problem still exists there please?

**Comment by stefanholek** [comment]
Let me give this a try...

```
openssl version
OpenSSL 3.3.1 4 Jun 2024 (Library: OpenSSL 3.3.1 4 Jun 2024)
```

Given the config file from above (server.conf), I attempt to create a CSR with:

```
openssl req -new -config server.conf -out foo.csr -keyout foo.key -addext subjectAltName=DNS:www.foo.com

Error adding request extensions defined via -addext
0056151501000000:error:0580008C:x509 certificate routines:X509at_add1_attr_by_NID:duplicate attribute:crypto/x509/x509_att.c:194:
```

The error is misleading as there is no "duplicate" attribute.

The only way to make this work is by removing the `[server_reqext]` section and adding all extensions via `-addext`. I was really hoping that `req_extensions` and `-addext` could be used in combination. 😞

So, @nhorman, `-addext` in 3.3 no longer wipes existing extensions but instead refuses to work at all if any extension has been set in the config file.

**Comment by t8m** [comment]
@DDvO was trying to improve the handling of extensions in req and other commands. This sounds like there are some issues remaining.

**Comment by DDvO** [comment]
> @DDvO was trying to improve the handling of extensions in req and other commands. 
> This sounds like there are some issues remaining.

Right. 

I meanwhile found that this newly reported issue dates back to the very introduction of the `-addext` option, which was in #4986.
There no care was taken that this option does not combine with other options that also add extensions.
The simple reason is that PKCS#10 CSR must have only one attribute providing X.509v3 extensions,
and the function adding such a list, namely `X509_REQ_add_extensions()`, must not be called more than once on the same req because it attempts to add a new such attribute.

There are essentially two ways out:
* combine beforehand the extensions to add to such a req
* make `X509_REQ_add_extensions()` more flexible 
by appending the given extensions to the pre-existing ones in case the req already includes an extension list.

**Comment by t8m** [comment]
This will be fixed by the means of #24792 in 3.4 version.
